src/snowflake/connector/config_parser/parser.py

```diff
@@ -0,0 +1,188 @@
+#
+# Copyright (c) 2012-2023 Snowflake Computing Inc. All rights reserved.
+#
+
+from __future__ import annotations
+
+import logging
+import os
+from collections.abc import Iterable
+from functools import wraps
+from pathlib import Path
+from typing import Callable, Literal, TypeVar
+
+import tomlkit
+from tomlkit import TOMLDocument
+from tomlkit.items import Table
+
+from ..errors import ConfigParserError, ConfigSourceError
+
+_T = TypeVar("_T")
+
+LOGGER = logging.getLogger(__name__)
+
+
+class ConfigOption:
+    def __init__(
+        self,
+        name: str,
+        _root_parser: ConfigParser,
+        _nest_path: list[str],
+        _type: Callable[[str], _T] | None = None,
+        choices: Iterable[_T] | None = None,
+        env_name: str | None | Literal[False] = None,
+    ) -> None:
+        """Create a config option that can read values from different locations.
+
+        Args:
+            name: The name of the ConfigOption
+            env_name: Environmental variable value should be read from, if not supplied, we'll construct this
+            type: A function that can turn str to the desired type, useful for reading value from environmental variable
+        """
+        self.name = name
+        self.type = _type
+        self.choices = choices
+        self._nest_path = _nest_path + [name]
+        self._root_parser: ConfigParser = _root_parser
+        self.env_name = env_name
+
+    def get(self):
```

Comment:
I am mostly a fan of `@property` instead of `get()` method approach:

```python
@property
def value(self):
    ....

@value.setter
def value(self):
    ....
````

Reply:
I don't like property here, because if you're debugging this code in PyCharm (for example) it'll continuously evaluate it.
What's worse is that might trigger reading your config file in and cache it. I want this action in specific to be done on purpose. Therefore; I went with the `get` function to signify that this has some cost and I hope you saved your config file previously.

Reply:
Can we add docstrings for all methods here?

Reply:
> I don't like property here, because if you're debugging this code in PyCharm (for example) it'll continuously evaluate it.

I don't understand this comment.  What gets continuously evaluated?